app/queries/get_change_offer_options.rb

```diff
@@ -0,0 +1,83 @@
+class GetChangeOfferOptions
+  attr_accessor :user, :accredited_provider, :recruitment_cycle_year
+
+  def initialize(user:, current_course:)
```

Comment:
In order to be able to implement the change link for Changing a course, we expect to be able to initialise the service with the selected provider, and be able to get back the list of courses.

Reply:
This service focuses on getting the list of `offerable_courses` for a user, which is a function of the current user and a course context (for preserving the ratifying provider). In this sense it's independent of any choices the user makes in the wizard.

For getting the list of courses for a given provider you could do `.offerable_courses.where(provider: ...)`, as all public methods in this service return active record queries. Likewise, for verifying the user has make_decisions on the chosen course, you carry on scoping e.g. `.offerable_courses.find_by(id: course_id)`

Once you reach the course level, and you verify the course is appropriate for this user, you don't necessarily need to scope anymore.

We could add a `def courses(provider:)` to the service, if it helps.

Reply:
Did we miss a card then? We were expecting to get a service that we would just plug in, and it would give us back anything we needed depending on what we requested from it. Also the card name is a bit deceiving then `Add new permissions-based query service for courses and **providers**`

Reply:
I don't think so. As far as I know, these are the queries you'll need:

- A list of training providers based on courses the user has make_decisions on, already available via `.available_providers`. This needs the context of a user and a ratifying provider, which is extracted from a course (e.g. application_choice.offered_course). This list is not influenced by any choices the user makes in the wizard
- A list of courses based on the provider the user has chosen in the wizard. This needs to be a subset of courses the user has make_decisions on, not all courses the chosen provider runs. Therefore, it's `offerable_courses` (or just `.courses`, if you prefer), scoped to a provider id.
- A list of study modes for the chosen course, available via either `.study_mode` on the selected course or by plucking `.study_mode` from the course's course_options.
- A list of locations for the chosen course, available via plucking `site` from the course's course_options.

Do we really need another service for the last queries? If it helps, we can wrap these in their own methods. Let's discuss on Monday.

Reply:
See also https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/app/queries/get_all_change_options_from_offered_option.rb#L21

Reply:
In line with our earlier chat, we now have `available_courses`, `available_study_modes`, `available_sites` and `available_course_options` as well as `available_providers`, but these new methods require additional context which is only relevant as you progress in the wizard.

I've added both `available_sites` and `available_course_options`, it's up to you which one you use during the 'location' part of the journey (using course options instead of sites for location radio buttons would allow you to pass a course_option_id, instead of a site_id, with the POST).